### PR TITLE
feat(cli): add typegen command for dynamic ReChunkEntries generation

### LINKED ISSUE #63: importChunk strict types
**Is your feature request related to a problem? Please describe.**
Currently importChunk accepts a string - so there is no strict validation that the chunk actually exists in `rechunk.json` - this would only be seen as an issue at runtime.

**Describe the solution you'd like**
Create a generate types command that will update the index.d.ts file with jscodeshift package and additionally add a typescript error when validation fails.

**Describe alternatives you've considered**
Keep as is.

### PR DESCRIPTION
**Pull request by crherman7**
## Scope

[closes TICKET-63](https://github.com/crherman7/rechunk/issues/63)

**Note** Merge after https://github.com/crherman7/rechunk/pull/62

## Implementation

1. **New `typegen` Command:**
   - Registers a `typegen` command in the CLI using the `commander` library.
   - Updates the `ReChunkEntries` type in the `index.d.ts` file dynamically based on `rechunk.json` entries.
   - Ensures the `importChunk` function only accepts valid chunk IDs.

2. **Enhanced TypeScript Type Definitions:**
   - Introduced `ValidateReChunkEntry<T>` to validate chunk IDs at the type level.
   - Improved error messages in `ValidateReChunkEntry<T>` with actionable instructions for regenerating types:
     ```typescript
     "Invalid ReChunk entry: '${T}' is not a recognized chunk ID. Please ensure the chunk ID exists 
     in the rechunk.json configuration file and regenerate types using the `typegen` command."
     ```

3. **`withTS` Utility for AST Transformations:**
   - Added a new `withTS` function to programmatically manipulate TypeScript AST using the TypeScript Compiler API.
   - Applied `withTS` in `typegen` to find and update the `ReChunkEntries` type.

4. **Improved Developer Experience:**
   - Added meaningful log messages with emojis for clarity and engagement during CLI execution.
   - Provided comprehensive TSDocs for `typegen`, `withTS`, and updated types, ensuring maintainability and ease of use.

## Screenshots

<details>
  <summary>editor</summary>

https://github.com/user-attachments/assets/dbd3a2d3-e726-4722-8e3f-8dc99b68d808
</details>

## How to Test

- Manually tested the typegen command to ensure:
        - It correctly updates the ReChunkEntries type based on rechunk.json.
        - It produces meaningful error messages for invalid chunk IDs.
- Verified AST manipulation through withTS for various scenarios.

## Emoji Guide

**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**

E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.

> 🟢 Nice refactor!

> 🟡 Why was the default value removed?

E.g: Blocking feedback must be addressed before merging.

> 🔴 This change will break something important

| | | |
| --- | --- | --- |
| Blocking | 🔴 ❌ 🚨 | RED |
| Non-blocking | 🟡 💡 🤔 💭 | Yellow, thinking, etc |
| Praise | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |

### CODE REVIEW

File: packages/cli/src/cmds/typegen.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
+import {program} from 'commander';
+import path from 'path';
```
Comment (devs9) [inline packages/cli/src/cmds/typegen.ts]: 🟡 is not used

File: packages/cli/src/cmds/typegen.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+              node.typeParameters,
+              ts.factory.createUnionTypeNode(
+                Object.keys(rc.entry).map(entry =>
+                  ts.factory.createLiteralTypeNode(
+                    ts.factory.createStringLiteral(entry),
+                  ),
+                ),
+              ),
+            );
+          }
```
Comment (devs9) [inline packages/cli/src/cmds/typegen.ts:96]: the dynamic updating types look awesome (with `withTS`), great job 😍

File: packages/cli/src/cmds/typegen.ts (hunk 1 above)
Comment (crherman7) [reply packages/cli/src/cmds/typegen.ts]: great catch!

**Review by devs9 — APPROVED** [review]
Awesome job 💚

**Pull request merged**